util/search_geo_arrayexpress.py

```diff
@@ -0,0 +1,119 @@
+from datetime import datetime
+import requests
+import xmltodict
+
+###############################################################################
+# This script queries two databases (GEO, ArrayExpress) to find human samples
+# analyzed on array or RNA-seq platforms. It parses information from each data
+# set and adds up the total number of samples from each platform. The output is
+# table showing the number of samples from each database and platform, plus a
+# ratio of the number of array samples to the number of RNA-seq samples. Also,
+# since this information will change over time, a timestamp representing the
+# moment these queries took place is included in the output.
+#
+# Usage: python3 util/search_geo_arrayexpress.py > output_file
+#
+# S. Foltz February 2022
+###############################################################################
+
+###############################################################################
+# GEO - Gene Expression Omnibus
+###############################################################################
+
+# set up search terms and dictionary to track n_samples
+search_base = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?"
+geo_array_search_term = "homo+sapiens[Organism]+AND+expression+profiling+by+array[DataSet+Type]"
+geo_rnaseq_search_term = "homo+sapiens[Organism]+AND+expression+profiling+by+high+throughput+sequencing[DataSet+Type]"
+
+geo_array_initial_url = search_base +
+"&".join(["db=gds", "term=" + geo_array_search_term])
+geo_rnaseq_initial_url = search_base +
+"&".join(["db=gds", "term=" + geo_rnaseq_search_term])
+
+geo_dict = {"array": [geo_array_initial_url, 0],
+            "rnaseq": [geo_rnaseq_initial_url, 0]}
+
+# for each platform in the dictionary, search twice and then fetch samples
+for platform in geo_dict:
+
+    # first search to retrieve the total number of results
+    initial_url = geo_dict[platform][0]
+    initial_xml = requests.get(initial_url)
+    initial_dict = xmltodict.parse(initial_xml.content)
+    n_results = initial_dict['eSearchResult']['Count']
+
+    # search again using n_results, save query_key and WebEnv for fetch
+    second_url = initial_url + "&RetMax=" + n_results + "&usehistory=y"
+    second_xml = requests.get(second_url)
+    second_dict = xmltodict.parse(second_xml.content)
+    query_key = second_dict['eSearchResult']['QueryKey']
+    webenv = second_dict['eSearchResult']['WebEnv']
+
+    # now fetch the records of the search results
+    fetch_base = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?"
```

Comment:
I would suggest moving these variables `fetch_base` and `skip_these` to BEFORE the for loop, at the top of the script. They are not being re-defined in the loop, so this is inefficient. While efficiency isn't a totally an issue deal for a python script like this, it's generally better practice to not be redefining these variables to the same value at every loop.